src/equations/equations.jl

```diff
@@ -1,5 +1,70 @@
+@muladd begin
+#! format: noindent
+
 using Trixi: AbstractEquations
 
+# Physical constants
+const EARTH_RADIUS = 6.37122  # m
+const EARTH_GRAVITATIONAL_ACCELERATION = 9.80616  # m/s²
+const EARTH_ROTATION_RATE = 7.292e-5  # rad/s
+const SECONDS_PER_DAY = 8.64e4
+
+@doc raw"""
+    AbstractCovariantEquations{NDIMS, 
+                               NDIMS_AMBIENT, 
+                               NVARS} <: AbstractEquations{NDIMS, NVARS} 
+
+Abstract type used to dispatch on systems of equations in covariant form, in which fluxes
+and prognostic variables are stored and computed in terms of their contravariant components 
+defining their expansions in terms of the local covariant tangent basis. The type parameter
+`NDIMS` denotes the dimension of the manifold on which the equations are solved, while
+`NDIMS_AMBIENT` is the dimension of the ambient space in which such a manifold is embedded. 
+
+!!! note 
+    Components of vector-valued fields should be prescibed within the global coordinate 
+    system (i.e. zonal and meridional components in the case of a spherical shell). 
+    When dispatched on this type, the function `Trixi.compute_coefficients!` will 
+    internally use the `covariant_basis` field of the container type 
+    [`P4estElementContainerCovariant`](@ref) to obtain the local contravariant components
+    used in the solver. 
+"""
+abstract type AbstractCovariantEquations{NDIMS,
+                                         NDIMS_AMBIENT,
+                                         NVARS} <: AbstractEquations{NDIMS, NVARS} end
+
+# Numerical flux plus dissipation which passes node/element indices and cache. 
+# We assume that u_ll and u_rr have been transformed into the same local coordinate system.
+@inline function (numflux::Trixi.FluxPlusDissipation)(u_ll, u_rr,
+                                                      orientation_or_normal_direction,
+                                                      equations::AbstractCovariantEquations{2},
+                                                      elements, i_ll, j_ll, i_rr, j_rr,
```

Comment:
I think that it would be good to change the name of the variable `elements` to something more descriptive (e.g., `cache_elements`) here and in other parts of the code, because `elements` can be a bit confusing.
```suggestion
                                                      cache_elements, i_ll, j_ll, i_rr, j_rr,
```